src/core/ddsc/src/dds_writer.c

```diff
@@ -298,15 +298,50 @@ static bool dds_writer_support_shm(const struct ddsi_config* cfg, const dds_qos_
     return false;
   }
 
-  return (NULL != qos &&
-          DDS_WRITER_QOS_CHECK_FIELDS ==
-              (qos->present & DDS_WRITER_QOS_CHECK_FIELDS) &&
+  // only VOLATILE or TRANSIENT LOCAL
+  if(!(qos->durability.kind == DDS_DURABILITY_VOLATILE ||
+    qos->durability.kind == DDS_DURABILITY_TRANSIENT_LOCAL)) {
+    return false;
+  }
+
+  // only KEEP LAST
+  if(qos->history.kind != DDS_HISTORY_KEEP_LAST) {
+    return false;
+  }
+  // we cannot support the required history with iceoryx
+  if (qos->durability.kind == DDS_DURABILITY_TRANSIENT_LOCAL &&
+      qos->durability_service.history.kind == DDS_HISTORY_KEEP_LAST &&
+      qos->durability_service.history.depth >
+          (int32_t)iox_cfg_max_publisher_history()) {
+    return false;
+  }
+
+  return (DDS_WRITER_QOS_CHECK_FIELDS == (qos->present & DDS_WRITER_QOS_CHECK_FIELDS) &&
           DDS_LIVELINESS_AUTOMATIC == qos->liveliness.kind &&
-          DDS_INFINITY == qos->deadline.deadline &&
-          DDS_RELIABILITY_RELIABLE == qos->reliability.kind &&
-          DDS_DURABILITY_VOLATILE == qos->durability.kind &&
-          DDS_HISTORY_KEEP_LAST == qos->history.kind
-          );
+          DDS_INFINITY == qos->deadline.deadline);
+}
+
+static iox_pub_options_t create_iox_pub_options(const dds_qos_t* qos) {
+
+  iox_pub_options_t opts;
+  iox_pub_options_init(&opts);
+
+  if(qos->reliability.kind == DDS_RELIABILITY_RELIABLE) {
+    opts.subscriberTooSlowPolicy = ConsumerTooSlowPolicy_WAIT_FOR_CONSUMER;
+  }
+
+  if(qos->durability.kind == DDS_DURABILITY_VOLATILE) {
+    opts.historyCapacity = 0;
+  } else {
+    // Transient Local and stronger
+    if (qos->durability_service.history.kind == DDS_HISTORY_KEEP_LAST) {
```

Comment:
Not sure whether the logic is completely sound here. It is hard to test in https://github.com/eclipse-cyclonedds/cyclonedds-cxx/pull/238 as it seems it is not possible to create a DataWriter with non-default values for durability_service.history.kind (despite using `TopicQoS`).

I am unsure whether the ctor of `DataWriterQos` from `TopicQoS` there should even work in general, but apparently it compiles ... I did not really look into it.

Reply:
I think I know why that is: the spec doesn't list `DurabilityService` as a QoS on the Topic, but that must be an oversight because it is an essential part of the definition of "transient" behaviour. OpenSplice corrected that, but incorrectly dropped it from the DataWriter. Cyclone's C++ binding came from OpenSplice and I suppose nobody noticed.

What should work is setting the `DurabilityService` QoS on the topic, because it looks to me like that gets propagated by the C++ binding to the Cyclone core, and then the DataWriter QoS should then inherit from it. If it doesn't propagate, then I am tempted to call that a bug.

Reply:
If it the history depth is set on the topic QoS it appears to propagate correctly.